lib/publiccloud/instance.pm

```diff
@@ -350,55 +350,50 @@ sub wait_for_ssh {
+    if ($xaccess == 0) {
+        # helps in boottime tests,when hard reboot could change the instance address:
+        ssh_script_run("ssh-keyscan $args{public_ip} | tee -a ~/.ssh/known_hosts") if (get_var('PUBLIC_CLOUD_PERF_COLLECT') or get_var('PUBLIC_CLOUD_CHECK_BOOT_TIME'));
+        # check system is running:
+        $out = $self->ssh_script_output(cmd => 'sudo systemctl is-system-running --wait', proceed_on_failure => 1, username => $args{username});
+        if ($out =~ m/running/) {    # startup ok
+            $status = 'OK';
+        }
+        elsif ($out =~ m/degraded/) {    # failed services colletion
+            $status = $self->run_ssh_command(cmd => 'sudo systemctl --failed', proceed_on_failure => 1, username => $args{username});
         }
-        else {
-            # On boottime test we do hard reboot which may change the instance address
-            script_run("ssh-keyscan $args{public_ip} | tee -a ~/.ssh/known_hosts") if (get_var('PUBLIC_CLOUD_CHECK_BOOT_TIME'));
-
-            my $output = $self->run_ssh_command(
-                cmd => 'sudo journalctl -b | grep -E "Reached target (Cloud-init|Default|Main User Target)"',
-                proceed_on_failure => 1,
-                username => $args{username});
-            if ($output =~ m/Reached target.*/) {
-                return $duration;
-            }
-            elsif ($output =~ m/Permission denied \(publickey\).*/) {
-                die "ssh permission denied (pubkey)" unless $args{ignore_wrong_pubkey};
-            }
+        elsif ($out =~ m/maintenance|stopping|offline|unknown/) {    # system ko
+            $status = 'KO';
         }
-        sleep $sleep_period;
     }
-
-    script_run("ssh  -i /root/.ssh/id_rsa -v $args{username}\@$args{public_ip} true", timeout => 360);
-    # Debug output: We have occasional error in 'journalctl -b' - see poo#96464 - this will be removed soon.
-    # Exclude 'mr_test/saptune' test case as it will introduce random softreboot failures.
-    if (!get_var('PUBLIC_CLOUD_SLES4SAP')) {
-        $self->run_ssh_command(cmd => 'sudo journalctl -b', proceed_on_failure => 1, username => $args{username}, timeout => 360);
+    else {    # connection failed
+        $status = 'KO';
     }
-
-    unless ($args{proceed_on_failure}) {
-        my $error_msg;
-        if ($check_port) {
-            $error_msg = sprintf("Unable to reach SSH port of instance %s with public IP:%s within %d seconds", $self->{instance_id}, $self->{public_ip},
-                $args{timeout});
-        }
-        else {
-            $error_msg = sprintf("Can not reach systemd target on instance %s with public IP:%s within %d seconds",
-                $self->{instance_id}, $self->{public_ip}, $args{timeout});
-        }
-        croak($error_msg);
+    # result display
+    record_info("CHECK SSH", 'wait ssh ' . (time() - $start_time) . 's, System:' . $out . "\n" . $status);
```

Comment:
mmm I assume this line should hardly depend on what we have in `$out` and `$status`  , otherwise we will get `Use of uninitialized variable` perl's exceptions . e.g. when `$xaccess` is non-zero value `$out` will be undefined . I think we need to move this string generation inside `if($xaccess == 0){}else{}` and in each certain case generate different content for this string with use of different variables . e.g. : 
```
my $check_ssh_msg;
if($xaccess == 0)
{
......
 $check_ssh_msg = 'wait ssh ' . (time() - $start_time) . 's, System:' . $out . "\n" . $status;
}
else
{
 $check_ssh_msg = 'Failed to reach port'
}
record_info("CHECK SSH", $check_ssh_msg);

Reply:
I have updated the code in order to have always a coherent report

Reply:
please push this changes

Reply:
pushed a more fix. I also added `$info` var too, but as it is now, we could also use only `out` and `status`.